receive_buffer.go

```diff
@@ -0,0 +1,142 @@
+package connmux
+
+import (
+	"io"
+	"sync"
+	"time"
+)
+
+// receiveBuffer buffers incoming frames. It queues up available frames in a
+// channel and makes sure that those are read in order when filling reader's
+// buffers via the read() method. It also makes sure to send an ack whenever a
+// queued frame has been fully read.
+//
+// In order to bound memory usage, the channel holds only <windowSize> frames,
+// after which it starts back-pressuring. The sender knows not to send more
+// than <windowSize> frames so as to prevent this. Once the sender receives an
+// ACK from the receiver, it sends a subsequent frame and so on.
+type receiveBuffer struct {
+	ackFrame []byte
+	in       chan []byte
+	ack      chan []byte
+	pool     BufferPool
+	poolable []byte
+	current  []byte
+	closed   bool
+	mx       sync.RWMutex
+}
+
+func newReceiveBuffer(streamID []byte, ack chan []byte, pool BufferPool, windowSize int) *receiveBuffer {
+	// Make an ackFrame for this stream id
+	ackFrame := make([]byte, len(streamID))
+	copy(ackFrame, streamID)
+	setFrameType(ackFrame, frameTypeACK)
+
+	return &receiveBuffer{
+		ackFrame: ackFrame,
+		in:       make(chan []byte, windowSize),
+		ack:      ack,
+		pool:     pool,
+	}
+}
+
+// submit allows the session to submit a new frame to the receiveBuffer. If the
+// receiveBuffer has been closed, this is a noop.
+func (buf *receiveBuffer) submit(frame []byte) {
+	buf.mx.RLock()
+	closed := buf.closed
```

Comment:
closed status and close(buf.in) are linked, that's why we need to protect the whole block. Otherwise, the submit to buf.in on 52 can fail if buf.in has been closed.